server/src/main/java/org/elasticsearch/cluster/coordination/LeaderChecker.java

```diff
@@ -0,0 +1,268 @@
+            // In the PoC, the leader sent its current version to the follower in the response to a LeaderCheck, so the follower
+            // could detect if it was lagging. We'd prefer this to be implemented on the leader, so the response is just
+            // TransportResponse.Empty here.
+            transportService.sendRequest(leader, LEADER_CHECK_ACTION_NAME, new LeaderCheckRequest(transportService.getLocalNode()),
+                TransportRequestOptions.builder().withTimeout(leaderCheckTimeout).withType(Type.PING).build(),
+
+                new TransportResponseHandler<TransportResponse.Empty>() {
+                    @Override
+                    public void handleResponse(Empty response) {
+                        if (isClosed.get()) {
+                            logger.debug("closed check scheduler received a response, doing nothing");
+                            return;
+                        }
+
+                        failureCountSinceLastSuccess.set(0);
+                        scheduleNextWakeUp(); // logs trace message indicating success
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        if (isClosed.get()) {
+                            logger.debug("closed check scheduler received a response, doing nothing");
+                            return;
+                        }
+
+                        long failureCount = failureCountSinceLastSuccess.incrementAndGet();
+                        if (failureCount >= leaderCheckRetryCount) {
+                            logger.debug(new ParameterizedMessage("{} consecutive failures (limit [{}] is {}) so leader has failed",
+                                failureCount, LEADER_CHECK_RETRY_COUNT_SETTING.getKey(), leaderCheckRetryCount), exp);
+                            leaderFailed();
+                        } else {
+                            logger.debug(new ParameterizedMessage("{} consecutive failures (limit [{}] is {})",
+                                failureCount, LEADER_CHECK_RETRY_COUNT_SETTING.getKey(), leaderCheckRetryCount), exp);
+                            scheduleNextWakeUp();
+                        }
+                    }
+
+                    @Override
+                    public String executor() {
+                        return Names.GENERIC;
+                    }
+                });
+        }
+
+        private void leaderFailed() {
+            if (isClosed.compareAndSet(false, true)) {
+                onLeaderFailure.run();
+            } else {
+                logger.debug("already closed, not failing leader");
+            }
+        }
+
+        private void scheduleNextWakeUp() {
+            logger.trace("scheduling next check of {} for [{}] = {}", leader, LEADER_CHECK_INTERVAL_SETTING.getKey(), leaderCheckInterval);
+            transportService.getThreadPool().schedule(leaderCheckInterval, Names.GENERIC, new Runnable() {
```

Comment:
here, we can use ThreadPool.SAME as handleWakeUp is very lightweight

Reply:
Ok, I pushed 5051d1b37f845b6193570848b589c832acbb910b